doc/charm-structure.rst

```diff
@@ -0,0 +1,67 @@
+Charm Directory and File Structure
+==================================
+
+A typical charm would have the following parts:
+
+* ``config.yaml`` and ``metadata.yaml``;
+* ``hooks`` directory with **symlinks corresponding to event names**. The
+  symlinks point to a file to a python file ``run_main.py`` that:
+  * extends python module search path to include the ``lib`` directory;
+  * calls an entrypoint function from ``entrypoint.py``;
+* ``lib/juju/framework.py`` - contains the charm framework code;
+* ``lib/juju/entrypoint.py`` contains the code needed to:
+  * **initialize** the charm framework for a given hook invocation;
+  * **create** the **in-memory state** of the charm
+    * in-memory state includes observers, emitters, persisted local state
+      (handled by the framework) and Juju model state snapshot parts they care
+      about (the **world view**);
+    * a module-level ``Charm`` class definition from ``lib/charm.py`` provided
+      by a charm author is used.
+  * **re-emit** any **deferred** events;
+  * **emit** the **new** Juju-originating event for the current hook invocation;
+  * **commit** (persist) the framework state.
```

Comment:
As a side note, this is also "reimplementing" the entrypoint file in form of docs. In general we don't need to repeat in docs or inline comments what the code is already saying by itself. Instead, good docs describe the things that the code doesn't by itself, such as why we need to do something, or a summary of a complex operation that is about to take place.

Reply:
I think a good textual description of the core execution logic would be valuable, especially for somebody new to Juju and charms who would like to delve deeper into the implementation aspects. However, we could also just point readers to the implementation given that it is not very large.